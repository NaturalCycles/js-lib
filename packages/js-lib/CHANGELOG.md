See [Github Releases](https://github.com/NaturalCycles/js-lib/releases)
