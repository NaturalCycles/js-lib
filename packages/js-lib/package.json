{
  "name": "@naturalcycles/js-lib",
  "type": "module",
  "version": "15.0.0",
  "scripts": {
    "build": "dev-lib build",
    "test": "dev-lib test",
    "lint": "dev-lib lint",
    "bt": "dev-lib bt",
    "lbt": "dev-lib lbt",
    "test-tz1": "TZ=Europe/Stockholm yarn test local",
    "test-tz2": "TZ=JST-9 yarn test local",
    "test-ny": "TZ=GMT-0500 yarn test localTime"
  },
  "dependencies": {
    "tslib": "^2",
    "zod": "^3"
  },
  "devDependencies": {
    "@naturalcycles/dev-lib": "*",
    "@types/crypto-js": "^4",
    "@types/node": "^22",
    "@types/semver": "^7",
    "crypto-js": "^4",
    "dayjs": "^1"
  },
  "files": [
    "dist",
    "cfg",
    "src",
    "!src/test",
    "!src/**/*.test.*",
    "!src/**/__snapshots__"
  ],
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "engines": {
    "node": ">=22.12.0"
  },
  "publishConfig": {
    "provenance": true,
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git@github.com:NaturalCycles/js-libs.git",
    "directory": "packages/js-lib"
  },
  "description": "Standard library for universal (browser + Node.js) javascript",
  "keywords": [
    "js-lib",
    "lodash",
    "p-map",
    "standard library"
  ],
  "author": "Natural Cycles Team",
  "license": "MIT"
}
