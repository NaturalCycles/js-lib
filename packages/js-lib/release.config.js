import sharedConfig from '../../release.config.js'

export default sharedConfig
