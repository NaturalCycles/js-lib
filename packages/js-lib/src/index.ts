export * from './abort.js'
export * from './array/array.util.js'
export * from './array/range.js'
export * from './bot.js'
export * from './browser/adminService.js'
export * from './browser/analytics.util.js'
export * from './browser/i18n/fetchTranslationLoader.js'
export * from './browser/i18n/translation.service.js'
export * from './browser/imageFitter.js'
export * from './browser/script.util.js'
export * from './browser/topbar.js'
export * from './datetime/dateInterval.js'
export * from './datetime/localDate.js'
export * from './datetime/localTime.js'
export * from './datetime/timeInterval.js'
export * from './datetime/wallTime.js'
export * from './decorators/asyncMemo.decorator.js'
export * from './decorators/createPromiseDecorator.js'
export * from './decorators/debounce.decorator.js'
export * from './decorators/debounce.js'
export * from './decorators/decorator.util.js'
export * from './decorators/logMethod.decorator.js'
export * from './decorators/memo.decorator.js'
export * from './decorators/memo.util.js'
export * from './decorators/memoFn.js'
export * from './decorators/memoFnAsync.js'
export * from './decorators/retry.decorator.js'
export * from './decorators/timeout.decorator.js'
export * from './define.js'
export * from './deviceIdService.js'
export * from './enum.util.js'
export * from './env.js'
export * from './env/buildInfo.js'
export * from './error/assert.js'
export * from './error/error.model.js'
export * from './error/error.util.js'
export * from './error/errorMode.js'
export * from './error/try.js'
export * from './error/tryCatch.js'
export * from './form.util.js'
export * from './http/fetcher.js'
export * from './http/fetcher.model.js'
export * from './http/http.model.js'
export * from './is.util.js'
export * from './iter/asyncIterable2.js'
export * from './iter/iterable2.js'
export * from './json-schema/from-data/generateJsonSchemaFromData.js'
export * from './json-schema/jsonSchema.cnst.js'
export * from './json-schema/jsonSchema.model.js'
export * from './json-schema/jsonSchema.util.js'
export * from './json-schema/jsonSchemaBuilder.js'
export * from './log/commonLogger.js'
export * from './math/accumulatingAverage.js'
export * from './math/math.util.js'
export * from './math/sma.js'
export * from './math/stack.util.js'
export * from './nanoid.js'
export * from './number/createDeterministicRandom.js'
export * from './number/number.util.js'
export * from './object/deepEquals.js'
export * from './object/map2.js'
export * from './object/object.util.js'
export * from './object/set2.js'
export * from './object/sortObject.js'
export * from './object/sortObjectDeep.js'
export * from './polyfill.js'
export * from './promise/abortable.js'
export * from './promise/pDefer.js'
export * from './promise/pDelay.js'
export * from './promise/pFilter.js'
export * from './promise/pHang.js'
export * from './promise/pMap.js'
export * from './promise/pProps.js'
export * from './promise/pQueue.js'
export * from './promise/pRetry.js'
export * from './promise/pState.js'
export * from './promise/pTimeout.js'
export * from './semver.js'
export * from './string/case.js'
export * from './string/escape.js'
export * from './string/hash.util.js'
export * from './string/json.util.js'
export * from './string/leven.js'
export * from './string/pupa.js'
export * from './string/readingTime.js'
export * from './string/regex.js'
export * from './string/safeJsonStringify.js'
export * from './string/slugify.js'
export * from './string/string.util.js'
export * from './string/stringify.js'
export * from './string/url.util.js'
export * from './time/time.util.js'
export * from './typeFest.js'
export * from './types.js'
export * from './unit/size.util.js'
export * from './web.js'
