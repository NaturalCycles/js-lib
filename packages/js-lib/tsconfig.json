{
  "extends": "@naturalcycles/dev-lib/cfg/tsconfig.json",
  "compilerOptions": {
    "target": "es2023",
    "lib": ["esnext", "dom", "dom.iterable"],
  },
  "include": ["src"],
  "exclude": ["**/__exclude"],
}
