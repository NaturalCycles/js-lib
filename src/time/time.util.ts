import type { NumberOfMilliseconds, UnixTimestampMillis } from '../types'

/**
 * using _ = blockTimer()
 * // will log "took 1.234 sec" on dispose
 *
 * using _ = blockTimer('named')
 * // will log "named took 1.234 sec" on dispose
 *
 * @experimental
 */
export function _blockTimer(name?: string): Disposable {
  const started = Date.now() as UnixTimestampMillis
  return {
    [Symbol.dispose](): void {
      console.log(`${name ? name + ' ' : ''}took ${_since(started)}`)
    },
  } as any
}

/**
 * Returns time passed since `from` until `until` (default to Date.now())
 */
export function _since(
  from: UnixTimestampMillis,
  until: UnixTimestampMillis = Date.now() as UnixTimestampMillis,
): string {
  return _ms(until - from)
}

/**
 * Returns, e.g:
 * 125 ms
 * 1.125 sec
 * 11 sec
 * 1m12s
 * 59m2s
 * 1h3m12s
 */
export function _ms(millis: NumberOfMilliseconds): string {
  // <1 sec
  if (millis < 1000) return `${Math.round(millis)} ms`

  // < 5 sec
  if (millis < 5000) {
    const s = millis / 1000
    return `${Math.trunc(s) === s ? s : s.toFixed(3)} sec`
  }

  const sec = Math.floor(millis / 1000) % 60
  const min = Math.floor(millis / (60 * 1000)) % 60
  const hrs = Math.floor(millis / (3600 * 1000))

  // <1 hr
  if (hrs === 0) {
    // <1 min
    if (min === 0) return `${sec} sec`

    return `${min}m${sec}s`
  }

  if (hrs < 24) {
    return `${hrs}h${min}m`
  }

  if (hrs < 48) {
    return `${Math.round(hrs + min / 60)}h`
  }

  // >= 48 hours

  const days = Math.floor(hrs / 24)
  return `${days} days`
}
