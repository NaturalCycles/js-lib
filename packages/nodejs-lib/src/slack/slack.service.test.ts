import { mockTime } from '@naturalcycles/dev-lib/dist/testing/time.js'
import { _stringify, commonLoggerNoop, Fetcher, pExpectedError } from '@naturalcycles/js-lib'
import { beforeEach, expect, test, vi } from 'vitest'
import { slackDefaultMessagePrefixHook, SlackService } from './slack.service.js'

let slackService = new SlackService({
  webhookUrl: 'https://dummyhook.com',
  logger: commonLoggerNoop,
  // defaults: {
  //   channel: 'test',
  // },
})

let lastBody: any

beforeEach(() => {
  lastBody = null
  mockTime()

  vi.spyOn(Fetcher, 'callNativeFetch').mockImplementation(async (url, init) => {
    lastBody = init.body

    if (url.includes('error')) {
      return new Response('some error!', {
        status: 500,
      })
    }

    return new Response('ok')
  })
})

test('basic test', async () => {
  await slackService.log('hey')
  expect(lastBody).toMatchSnapshot()

  await slackService.send('hey')
  expect(lastBody).toMatchSnapshot()

  // Array of 1 item
  await slackService.send({ items: ['yo'] })
  expect(lastBody).toMatchSnapshot()

  // Just adding "everything" for coverage/snapshot
  const err = new Error('c error!')
  delete err.stack // to not include stack in the tested snapshot (flaky)

  await slackService.send(
    {
      items: ['a', { b: 'b' }, err],
      kv: { k1: 'v1' },
      mentions: ['kirill'],
    },
    { ctxKey: 'ctxValue' },
  )
  expect(lastBody).toMatchSnapshot()

  // await expect(slackService.send({items: 'hey', throwOnError: true})).rejects.toThrow('Network request forbidden')
})

test('no webhookUrl', async () => {
  slackService = new SlackService({
    logger: commonLoggerNoop,
  })
  await slackService.log('anything')
  expect(lastBody).toBeNull() // should be unchanged
})

test('error', async () => {
  slackService = new SlackService({
    webhookUrl: 'wrongUrl',
    logger: commonLoggerNoop,
  })

  // This should NOT throw, because errors are suppressed
  await slackService.log('yo')

  // this unmocks Fetcher and lets Slack fail
  vi.restoreAllMocks()

  const err = await pExpectedError(slackService.send({ items: 'yo', throwOnError: true }))
  expect(_stringify(err)).toMatchInlineSnapshot(`
    "HttpRequestError: POST wrongUrl
    Caused by: TypeError: Failed to parse URL from wrongUrl
    Caused by: TypeError: Invalid URL"
  `)
})

test('messagePrefixHook returning null should NOT be sent', async () => {
  slackService = new SlackService({
    webhookUrl: 'https://valid.com',
    messagePrefixHook: () => null,
    logger: commonLoggerNoop,
  })

  await slackService.log('yo')
  expect(lastBody).toBeNull()

  // revert to default hook
  slackService.cfg.messagePrefixHook = slackDefaultMessagePrefixHook
  await slackService.log('yo')
  expect(lastBody).not.toBeNull()
})
