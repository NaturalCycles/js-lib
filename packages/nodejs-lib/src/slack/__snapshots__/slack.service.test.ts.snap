// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic test 1`] = `"{"username":"bot","channel":"#log","icon_emoji":":spider_web:","text":"2018-06-21 00:00:00\\nhey"}"`;

exports[`basic test 2`] = `"{"username":"bot","channel":"#log","icon_emoji":":spider_web:","text":"2018-06-21 00:00:00\\nhey"}"`;

exports[`basic test 3`] = `"{"username":"bot","channel":"#log","icon_emoji":":spider_web:","text":"2018-06-21 00:00:00\\n\`\`\`yo\`\`\`"}"`;

exports[`basic test 4`] = `"{"username":"bot","channel":"#log","icon_emoji":":spider_web:","mentions":["kirill"],"attachments":[{"fields":[{"title":"k1","value":"v1","short":true}]}],"text":"2018-06-21 00:00:00\\n\`\`\`a\\n{ b: 'b' }\\nError: c error!\`\`\`\\n<@kirill>"}"`;
